File: Server/src/main/java/jk_5/nailed/server/command/CommandTP.java
package jk_5.nailed.server.command;

import jk_5.nailed.map.Map;
import jk_5.nailed.map.MapLoader;
import jk_5.nailed.map.mappack.Spawnpoint;
import jk_5.nailed.map.teleport.TeleportHelper;
import jk_5.nailed.map.teleport.TeleportOptions;
import jk_5.nailed.players.Player;
import net.minecraft.command.CommandException;
import net.minecraft.command.ICommandSender;
import net.minecraft.command.PlayerNotFoundException;
import net.minecraft.command.WrongUsageException;
import net.minecraft.entity.Entity;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.ChatComponentTranslation;

import java.util.List;

/**
 * No description given
 *
 * @author jk-5
 */
public class CommandTP extends NailedCommand {

    @Override
    public String getCommandName(){
        return "tp";
    }

    @Override
    public int getRequiredPermissionLevel(){
        return 2;
    }

    @Override
    public List addTabCompletionOptions(ICommandSender par1ICommandSender, String[] args){
        if(args.length == 1) return getListOfStringsMatchingLastWord(args, MinecraftServer.getServer().getConfigurationManager().getAllUsernames());
        if(args.length == 2) return getListOfStringsMatchingLastWord(args, MinecraftServer.getServer().getConfigurationManager().getAllUsernames());
        return null;
    }

    @Override
    public void processCommandPlayer(Player sender, Map map, String[] args){
        super.processCommandPlayer(sender, map, args);
    }

    @Override
    public void process(ICommandSender sender, String[] args){
        Entity subject;
        TeleportOptions link;
        try{
            String sSubject = null;
            String sTarget;
            String sX = null;
            String sY = null;
            String sZ = null;

            if(args.length > 3){
                if(args.length > 4){
                    sSubject = args[(args.length - 5)];
                }
                sTarget = args[(args.length - 4)];
                sX = args[(args.length - 3)];
                sY = args[(args.length - 2)];
                sZ = args[(args.length - 1)];
            }else if(args.length == 2){
                sSubject = args[(args.length - 2)];
                sTarget = args[(args.length - 1)];
            }else if(args.length == 1){
                sTarget = args[(args.length - 1)];
            }else{
                throw new WrongUsageException("commands.nailed.tp.usage");
            }

            if(sSubject == null)
                subject = getCommandSenderAsPlayer(sender);
            else{
                subject = getTargetPlayer(sender, sSubject);
            }
            if(subject == null){
                throw new WrongUsageException("commands.nailed.tp.fail.nosubject", new Object[0]);
            }

            link = getLinkInfoForTarget(sender, subject, sTarget, sX, sY, sZ);
            TeleportHelper.travelEntity(subject.worldObj, subject, link);
        }catch(CommandException e){
            sender.func_145747_a(new ChatComponentTranslation(e.getMessage()));
            sender.func_145747_a(new ChatComponentTranslation(getCommandUsage(sender)));
        }
    }

    public static TeleportOptions getLinkInfoForTarget(ICommandSender sender, Entity subject, String sTarget, String sX, String sY, String sZ){
        TeleportOptions link = null;
        try{
            Entity target = getTargetPlayer(sender, sTarget);
            link = createOptionsForLocation(new Spawnpoint(target), MapLoader.instance().getMap(target.worldObj));
        }catch(PlayerNotFoundException e){
            //NOOP. We'll try again later
        }
        if(link == null){
            link = new TeleportOptions();
            int dim = (int) (handleRelativeNumber(sender, subject.dimension, sTarget, 0, 0) - 0.5D);
            Map destMap = MapLoader.instance().getMap(dim);
            if(destMap == null){
                throw new CommandException("commands.nailed.tp.fail.noworld", dim);
            }
            link.setDestination(destMap);
            if((sX != null) && (sY != null) && (sZ != null)){
                int x = (int) handleRelativeNumber(sender, subject.posX, sX);
                int y = (int) handleRelativeNumber(sender, subject.posY, sY, 0, 0);
                int z = (int) handleRelativeNumber(sender, subject.posZ, sZ);
                link.setCoordinates(new Spawnpoint(x, y, z));
            }
        }
        return link;
    }

    private static TeleportOptions createOptionsForLocation(Spawnpoint spawnpoint, Map map){
        TeleportOptions options = new TeleportOptions();
        options.setCoordinates(spawnpoint);
        options.setDestination(map);
        return options;
    }
}

File: Server/src/main/scala/jk_5/nailed/server/command/CommandTp.scala
package jk_5.nailed.server.command

import net.minecraft.command.ICommandSender
import jk_5.nailed.api.map.Map
import scala.collection.mutable
import jk_5.nailed.api.map.teleport.TeleportOptions
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import jk_5.nailed.api.NailedAPI
import jk_5.nailed.map.Location

/**
 * No description given
 *
 * @author jk-5
 */
object CommandTp extends ScalaCommand {

  val name = "tp"
  val usage = "/tp <teleportingPlayer> [targetPlayer|x y z] - Teleport a player to another player or coordinates"

  override def processCommandWithMap(sender: ICommandSender, map: Map, args: Array[String]){
    val locations = mutable.ArrayBuffer[(Entity, TeleportOptions)]()
    args.length match {
      case 1 => //  /tp destination
        val teleporting = sender match {
          case p: EntityPlayer => p
          case _ => throw new CommandException("We can't teleport you as you aren't a player")
        }
        val target = Option(NailedAPI.getPlayerRegistry.getPlayerByUsername(args(0)))
        if(target.isEmpty) throw new CommandException(s"Player ${args(0)} was not found")
        val option = TeleportOptions.builder().location(target.get.getLocation).destination(target.get.getCurrentMap).build
        locations += ((teleporting, option))
      case 2 => //  /tp teleportingPlayer destinationPlayer | destinationMap
        val option = getDestination(sender, args(1))
        getPlayersList(sender, args(0)).foreach(p => locations += ((p, option)))
      case 3 => //  /tp x y z
        val teleporting = sender match {
          case p: EntityPlayer => p
          case _ => throw new CommandException("We can't teleport you as you aren't a player")
        }
        val x = handleRelativeNumber(sender, teleporting.posX, args(0))
        val y = handleRelativeNumber(sender, teleporting.posY, args(1), 0, 0)
        val z = handleRelativeNumber(sender, teleporting.posZ, args(2))
        val option = TeleportOptions.builder().location(new Location(x, y, z, teleporting.rotationYaw, teleporting.rotationPitch)).destination(NailedAPI.getMapLoader.getMap(teleporting.worldObj)).build()
        locations += ((teleporting, option))
      case 4 => //  /tp player x y z
        getPlayersList(sender, args(0)).foreach(p => {
          val x = handleRelativeNumber(sender, p.posX, args(0))
          val y = handleRelativeNumber(sender, p.posY, args(1), 0, 0)
          val z = handleRelativeNumber(sender, p.posZ, args(2))
          val option = TeleportOptions.builder().location(new Location(x, y, z, p.rotationYaw, p.rotationPitch)).destination(NailedAPI.getMapLoader.getMap(p.worldObj)).build()
          locations += ((p, option))
        })
      case 5 => //  /tp player x y z options
        val opt = args(4)
        val particles = !opt.contains("noparticle")
        val sound = !opt.contains("nosound")
        val momentum = opt.contains("momentum")
        val clearInv = opt.contains("clearinventory")
        getPlayersList(sender, args(0)).foreach(p => {
          val x = handleRelativeNumber(sender, p.posX, args(0))
          val y = handleRelativeNumber(sender, p.posY, args(1), 0, 0)
          val z = handleRelativeNumber(sender, p.posZ, args(2))
          val option = new TeleportOptions
          option.setSound(if(sound) option.getSound else null)
          option.setSpawnParticles(particles)
          option.setMaintainMomentum(momentum)
          option.setClearInventory(clearInv)
          option.setLocation(new Location(x, y, z, p.rotationYaw, p.rotationPitch))
          option.setDestination(NailedAPI.getMapLoader.getMap(p.worldObj))
          locations += ((p, option))
        })
      case _ => throw new WrongUsageException("/tp <teleportingPlayer> [targetPlayer|x y z]")
    }
    locations.foreach(p => NailedAPI.getTeleporter.teleportEntity(p._1, p._2))
  }

  override def addAutocomplete(sender: ICommandSender, args: Array[String]): List[String] =
    if(args.length == 1 || args.length == 2) getUsernameOptions(args) else null

  private def getDestination(sender: ICommandSender, data: String): TeleportOptions = {
    val dest = new TeleportOptions
    Option(NailedAPI.getPlayerRegistry.getPlayerByUsername(data)) match {
      case Some(player) =>
        dest.setLocation(player.getLocation)
        dest.setDestination(player.getCurrentMap)
      case None =>
        Option(NailedAPI.getMapLoader.getMap(data)) match {
          case Some(map) =>
            if(map.getMappack != null) dest.setLocation(map.getMappack.getMappackMetadata.getSpawnPoint)
            dest.setDestination(map)
          case None => throw new CommandException("No map/player found matching that name")
        }
    }
    dest
  }
}
